### ISSUE TITLE
Blob storage building block

### ISSUE THREAD

**Issue by artursouza** [dapr/dapr#4804]
Today, we use binding for multiple components, without a consistent API spec but there are a subset of those components that behave like a blob storage. There is an opportunity to consolidate those into a defined API spec (new building block). The API would be simple to POST and GET to/from blob storage, offering portability.

**Comment by yaron2** [comment]
> This is different from state store where the API surface is expected to be small with no deviation from the generic behavior of getting and posting one object.

If the API for blobs is a subset of the API for state, I'm concerned that multiple implementations (3 - bindings, state stores, blobs) might be confusing, and developers will naturally go to the one that has the most API surface, ie. state.

In the likely event that developers will also want to delete and post multiple objects, we'll see the Blob API become an almost mirror reflection of the state API.

Thoughts?

**Comment by ItalyPaleAle** [comment]
I think the missing part in current state stores, to make them effective for managing blobs, is support for streaming. Our current APIs load everything in memory in multiple steps (application and sidecar both load the entire blob in memory), so this is not ideal for large blobs.

A suggestion I brought up to @berndverst recently was to add support for streams to state stores, when the underlying service and SDK support that. We'd have two version of the APIs to work with byte slices and streams separately, for example:

```go
// Uses byte slices -> data is in the Data property in the response object
Get(req *state.GetRequest) (*state.GetResponse, error)
// Uses streams -> in the response object, Data is nil
GetStream(req *state.GetRequest, out io.Writer) (*state.GetResponse, error)
// Uses byte slices -> data is in the Value property of the request object
Set(req *state.SetRequest) error
// Uses streams -> in the request object, the Value property is nil
SetStream(req *state.StateRequest, in io.Reader) error
// Alternatively, we could set a stream in the req object and not add a new signature
```

For components that support streams, methods would only use streams. If users pass a byte slice, then a reader is created from that.

For components that do not support streams, you can still use the `*Stream` methods. Internally, the method reads the data from a stream into a buffer.

**Comment by olitomlinson** [comment]
> If the API for blobs is a subset of the API for state, I'm concerned that multiple implementations (3 - bindings, state stores, blobs) might be confusing, and developers will naturally go to the one that has the most API surface, ie. state.

Completely agree with this.

I still struggle with there being a first-class State Management building block, but then also a myriad of bindings that do state operations as well... it just feel somewhat fragmented and IMO doesn't paint a solid story for Dapr and state management concerns.

Would it really hurt to change the guidelines on bindings so that any contribution are directed more to towards the development of State Management components rather than bindings? This IMO would help unify state management, and provide better direction on what bindings are for but equally what they are _not_ for.

Current dilemma : I sell outwards to my technical peers in my org on the virtue of the State Management component, to find that there is no S3 State Management component available. S3 is literally one of the most canonical examples of AWS infrastructure (alongside EC2) and there is no State Management component. Which IMO is a missed opportunity for an easy sell of State Management portability across AWS and Azure with blobs.

This sounds a touch doomsday/emotive, but I'm going to say it anyway -  Please don't introduce an identity crisis into State Management with Dapr. (_Personally, I think bindings are already in a bit of an identity crisis in Dapr, but thats a conversation for another day..._)

**Comment by berndverst** [comment]
@ItalyPaleAle the suggestion makes sense but instead first requires support from dapr/dapr maintainers.

Today the sidecar does not support streaming calls from app to sidecar and of course the SDKs will need to support this too. There is no point in supporting streaming calls from sidecar to service if there is no way to stream data to the sidecar itself. So lets have @dapr/maintainers-dapr address that first. Once we have that I'm all for finding a way to indicate which components support streaming and expose that as an option.

@olitomlinson
Re the S3 component -- anyone with interest in S3 can contribute this. An actual frequent user of S3 would be the best person to do so.

Dapr State Store is a special purpose opinionated way to save and retrieve key value pairs. Sometimes you need more control and that's why bindings exist (e.g. you need to store data in a different manner, or need to access special database features etc). The state store interface allows components to be swapped out underneath, but bindings don't provide this guarantee. For this very reason it is often impossible to add features to the state store components as only features with implementations in all state stores can be added.
Bindings are the most flexible components that need not adhere to any particular structure. You call it fragmentation - I call it versatility. I do not expect any structure among bindings. Perhaps what you are asking for is a new building block type but that need not come at the expense of bindings.

**Comment by ItalyPaleAle** [comment]
> Today the sidecar does not support streaming calls from app to sidecar and of course the SDKs will need to support this too. There is no point in supporting streaming calls from sidecar to service if there is no way to stream data to the sidecar itself. 

Yes, this needs to be coordinated with the runtime. If components do support streaming but the runtime doesn't, then I agree there's no much value added.

Thus said, I do think that support for streaming will be key at least for state stores. It's not practical to use the current state stores to store large (many MBs to GBs) blobs.

**Comment by berndverst** [comment]
> > Today the sidecar does not support streaming calls from app to sidecar and of course the SDKs will need to support this too. There is no point in supporting streaming calls from sidecar to service if there is no way to stream data to the sidecar itself.
> 
> Yes, this needs to be coordinated with the runtime. If components do support streaming but the runtime doesn't, then I agree there's no much value added.
> 
> Thus said, I do think that support for streaming will be key at least for state stores. It's not practical to use the current state stores to store large (many MBs to GBs) blobs.

I agree - that is because our state stores are intended to be Key Value stores only.

Maybe an object storage building block would make sense (blob could be one such flavor) and streaming support should be one of the criteria there.

**Comment by ItalyPaleAle** [comment]
> Maybe an object storage building block would make sense (blob could be one such flavor) and streaming support should be one of the criteria there.

Or object storage could could just be a feature of (certain) state stores. Now that we have the capabilities API, we can also expose which state store supports object storage (prerequisite is support for streaming) and which don't?

For example: Cosmos DB would not mark itself as "supporting object storage". Azure Storage would.

**Comment by yaron2** [comment]
@berndverst you are spot on with needing to have `daprd` support streaming capabilities on the API layer before this can extend to one or more building blocks. The moment we have streaming support, extending this into building blocks should be easy enough with an interface very similar to what @ItalyPaleAle suggested.

FYI, we recently added the ability for the S3 binding to read a file from the filesystem and upload it. So while not streaming enabled, this does go a long way in removing the boilerplate code and I/O network operation with sending the data over to Dapr. If the S3 Go SDK supports streaming uploads directly from a file, this can very easily be added also.

**Comment by yaron2** [comment]
> Or object storage could could just be a feature of (certain) state stores

I like this direction better.

**Comment by yaron2** [comment]
@artursouza can this now be closed in favor of https://github.com/dapr/dapr/issues/4934?

**Comment by olitomlinson** [comment]
Bump

**Comment by berndverst** [comment]
> Bump

Should be closed in favor of #4934 anyway

**Comment by artursouza** [comment]
Given our current move to create Document Store and remove Query API from State Store - it feels backwards to bundle yet another functionality into the state store. The blob store API is fundamentally different from state store as it does not offer any batching operation and the payload is as-is - without any type of structure. This will also allow streaming to be provided end-to-end.

**Comment by ItalyPaleAle** [comment]
As I got older, wiser, and more mature on Dapr, I am now leaning towards a separate building block too :) Streaming would work regardless, but the philosophical differences that @artursouza is highlighting do make sense.

**Comment by ItalyPaleAle** [comment]
Formalized the proposal into a new building block: dapr/proposals#18
